v3.1/reference/transactions/transaction-optimistic.md

```diff
@@ -0,0 +1,165 @@
+---
+title: 乐观事务模型
+category: reference
+aliases: ['/docs-cn/v3.1/reference/transactions/transaction-model/']
+---
+
+# 乐观事务模型
+
+TiDB 默认使用乐观事务模型。也就是说，在执行 `UPDATE`、`INSERT`、`DELETE` 等语句时，只有在提交过程中才会检查写写冲突，而不是像 MySQL 一样使用行锁来避免写写冲突。类似的，诸如 `GET_LOCK()` 和 `RELEASE_LOCK()` 等函数以及 `SELECT .. FOR UPDATE` 之类的语句在 TiDB 和 MySQL 中的执行方式并不相同。所以业务端在执行 SQL 语句后，需要注意检查 `COMMIT` 的返回值，即使执行时没有出错，`COMMIT` 的时候也可能会出错。
+
+本文介绍 TiDB 乐观事务的原理，以及相关特性。本文假定你对 [TiDB 的整体架构](/v3.1/architecture.md#tidb-整体架构)和 [Percolator](https://www.usenix.org/legacy/event/osdi10/tech/full_papers/Peng.pdf) 事务模型都有一定了解，相关核心概念如下：
+
+- [ACID](/v3.1/glossary.md#acid)
+- [事务](/v3.1/glossary.md#事务)
+- [乐观事务](/v3.1/glossary.md#乐观事务)
+- [显式事务/隐式事务](/v3.1/glossary.md#显式事务隐式事务)
+
+## 乐观事务原理
+
+TiDB 中事务使用两阶段提交，流程如下：
+
+![TiDB 中的两阶段提交](/media/2pc-in-tidb.png)
+
+1. 客户端开始一个事务。
+
+    TiDB 从 PD 获取一个全局唯一递增的版本号作为当前事务的开始版本号，这里定义为该事务的 `start_ts` 版本。
+
+2. 客户端发起读请求。
+
+    a. TiDB 从 PD 获取数据路由信息，即数据具体存在哪个 TiKV 节点上。
+
+    b. TiDB 从 TiKV 获取 `start_ts` 版本下对应的数据信息。
+
+3. 客户端发起写请求。
+
+    TiDB 校验写入数据是否符合一致性约束（如数据类型是否正确、是否符合唯一索引约束等）。**校验通过的数据将存放在内存里。**
+
+4. 客户端发起 commit。
+
+5. TiDB 开始两阶段提交，保证分布式事务的原子性，让数据真正落盘。
+
+    a. TiDB 从当前要写入的数据中选择一个 Key 作为当前事务的 Primary Key。
+
+    b. TiDB 从 PD 获取所有数据的写入路由信息，并将所有的 Key 按照所有的路由进行分类。
+
+    c. TiDB 并发地向所有涉及的 TiKV 发起 prewrite 请求。TiKV 收到 prewrite 数据后，检查数据版本信息是否存在冲突或已过期。符合条件的数据会被加锁。
+
+    d. TiDB 成功收到所有 prewrite 请求。
+
+    e. TiDB 向 PD 获取第二个全局唯一递增版本号，定义为本次事务的 `commit_ts`。
+
+    f. TiDB 向 Primary Key 所在 TiKV 发起第二阶段提交。TiKV 收到 commit 操作后，检查数据合法性，清理 prewrite 阶段留下的锁。
+
+    g. TiDB 收到 f 成功信息。
+
+6. TiDB 向客户端返回事务提交成功的信息。
+
+7. TiDB 异步清理本次事务遗留的锁信息。
+
+## 优缺点分析
+
+通过分析 TiDB 中事务的处理流程，可以发现 TiDB 事务有如下优点：
+
+* 实现原理简单，易于理解。
+* 基于单实例事务实现了跨节点事务。
+* 锁管理实现了去中心化。
+
+但 TiDB 事务也存在以下缺点：
+
+* 两阶段提交使网络交互增多。
+* 缺少一个中心化的版本管理服务。
+* 事务数据量过大时易导致内存暴涨。
+
+### 重试机制
+
+TiDB 中默认使用乐观事务模型，因而在高冲突率的场景中，事务很容易提交失败。而 MySQL 内部使用的是悲观事务模型，在执行 SQL 语句的过程中进行冲突检测，所以提交时很难出现异常。为了兼容 MySQL 的悲观事务行为，TiDB 提供了重试机制。当事务提交后，如果发现冲突，TiDB 内部重新执行包含写操作的 SQL 语句。你可以通过设置 `tidb_disable_txn_auto_retry` 和 `tidb_retry_limit` 开启自动重试：
+
+```toml
+# 用于设置是否禁用自动重试，默认不重试。
+tidb_disable_txn_auto_retry = on
+# 用来控制重试次数。只有自动重试启用时该参数才会生效。
+# 当 “tidb_retry_limit= 0” 时，也会禁用自动重试。
+tidb_retry_limit = 10
+```
+
+推荐通过以下两种方式进行参数设置：
+
+1. Session 级别设置：
+
+    {{< copyable "sql" >}}
+
+    ```sql
+    set @@tidb_disable_txn_auto_retry = off;
```

Comment:
Update this block to make it copy-friendly. Each SQL statement in one code block.